src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,936 @@
+struct WeightedViaNodePackedPath
+{
+    WeightedViaNode via;
+    PackedPath path;
+    std::vector<EdgeWeight> path_weights;
+};
+
+// Represents a high-detail unpacked path (s, .., via, .., t)
+// its total weight and the via node used to construct the path.
+struct WeightedViaNodeUnpackedPath
+{
+    WeightedViaNode via;
+    UnpackedNodes nodes;
+    UnpackedEdges edges;
+};
+
+// Filters candidates which are on not unique.
+// Returns an iterator to the uniquified range's new end.
+// Note: mutates the range in-place invalidating iterators.
+template <typename RandIt> RandIt filterViaCandidatesByUniqueNodeIds(RandIt first, RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    std::sort(first, last, [](auto lhs, auto rhs) { return lhs.node < rhs.node; });
+    return std::unique(first, last, [](auto lhs, auto rhs) { return lhs.node == rhs.node; });
+}
+
+// Filters candidates which are on un-important roads.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterViaCandidatesByRoadImportanceHeuristic(RandIt first, RandIt last, const Facade &facade)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    // Todo: the idea here is to filter out alternatives where the via candidate is not on a
+    // high-priority road. We should experiment if this is really needed or if the boundary
+    // nodes the mld search space gives us already provides us with reasonable via candidates.
+    //
+    // Implementation: we already have `RoadClassification` from guidance. We need to serialize
+    // it to disk and then in the facades read it in again providing `IsImportantRoad(NodeID)`.
+    // Note: serialize out bit vector keyed by node id with 0/1 <=> unimportant/important.
+    (void)first;
+    (void)last;
+    (void)facade;
+
+    return last;
+}
+
+// Scale the maximum allowed weight increase based on its magnitude:
+//  - Shortest path 10 minutes, alternative 13 minutes => Factor of 0.30 ok
+//  - Shortest path 10 hours, alternative 13 hours     => Factor of 0.30 unreasonable
+double scaledAtMostLongerByFactorBasedOnDuration(EdgeWeight duration)
+{
+    auto scaledAtMostLongerBy = kAtMostLongerBy;
+
+    // We only have generic weights here and no durations without unpacking.
+    // We also have restricted way penalties which are huge and will screw scaling here.
+    //
+    // Users can pass us generic weights not based on durations; we can't do anything about
+    // it here other than either generating too many or no alternatives in these cases.
+    //
+    // We scale the weights with a step function based on some rough guestimates, so that
+    // they match tens of minutes, in the low hours, tens of hours, etc.
+
+    const constexpr auto minutes = 60.;
```

Comment:
Should it be `60. * weight_multiplier` here? Also it will work only for time-based weights